### PR TITLE
Separate geometric interpolation space from function interpolation space

### PR DESCRIPTION
**Pull request by fredrikekre**
This PR separates the geometric interpolation from the function interpolation in `FEValues`. This makes it easier to extend to sub/super-parametric elements. It works quite nice, since the only extra thing that needs to be stored is the derivative of a (potential) other function space. Defaults to iso-parametric, so it doesn't break anything.

This also adds a new function space, `Lagrange{2,RefCube,2}` which were missing.

Some things we should decide before merging:

- [x] ~~What should the two function spaces be called in the `FEValues` type?~~
- `FunctionSpace (FS)` and `GeometricSpace (GS)`
- [x] ~~What symbol should be used for the geometric functions? Cf. `N` for the other function space. Currently I use `M`.~~
- `M`
- [x] ~~What should we call the `get` functions to get the function spaces? Previously `get_functionspace(fev::FEValues)` made more sense. I changed to `get_func_functionspace(fev::FEValues)` and `get_geom_functionspace(fev::FEValues)` to get the function space used for approximating the function and the space to interpolate the geometry respectively. It is kind of ugly, but the are mostly used internally anyways. Lets name them based on the decision of the first point above.~~
- `get_functionspace()` and `get_geometricspace()`

### CODE REVIEW

**Comment by fredrikekre** [comment]
I found one source where they call the geometric interpolation functions: `shape functions` and the functions used to interpolate the approximated field `interpolation functions`.

File: src/fe_values.jl [hunk 1]
```diff
@@ -38,25 +38,33 @@ immutable FEValues{dim, T <: Real, FS <: FunctionSpace}
     detJdV::Vector{T}
     quad_rule::QuadratureRule{dim, T}
     function_space::FS
+    dMdξ::Vector{Vector{Vec{dim, T}}}
+    geom_space::FS
```
Comment (KristofferC) [inline src/fe_values.jl]: If this is FS then function_space and geom_space has to be the same.

File: src/function_spaces.jl [hunk 2]
```diff
@@ -186,11 +181,80 @@ function reference_coordinates(fs::Lagrange{2, RefCube, 1})
...
 
+##################################
+# Lagrange dim 2 RefCube order 2 #
+##################################
+n_basefunctions(::Lagrange{2, RefCube, 2}) = 8
 
-###################################
-# Lagrange dim 2 RefTetrahedron order 1 #
-###################################
+function value!(fs::Lagrange{2, RefCube, 2}, N::Vector, ξ::Vec{2})
```
Comment (KristofferC) [inline src/function_spaces.jl]: Isn't this what I call serendipity below? A la grange element would have 9 nodes?

File: src/fe_values.jl [hunk 3]
```diff
@@ -38,25 +38,33 @@ immutable FEValues{dim, T <: Real, FS <: FunctionSpace}
...
 end
 
-FEValues{dim, FS <: FunctionSpace}(quad_rule::QuadratureRule{dim}, func_space::FS) = FEValues(Float64, quad_rule, func_space)
+FEValues{dim, FS <: FunctionSpace}(quad_rule::QuadratureRule{dim}, func_space::FS, geom_space::FS=func_space) = FEValues(Float64, quad_rule, func_space, geom_space)
```
Comment (fredrikekre) [inline src/fe_values.jl]: Okay, is it enough to have like:
`FEValues{dim, FS1 <: FunctionsSpace, FS2 <: FunctionSpace}(...)`

File: src/function_spaces.jl (hunk 2 above)
Comment (fredrikekre) [inline src/function_spaces.jl]: Oh.... Will update later

Comment (fredrikekre) [inline src/function_spaces.jl]: I got confused from the VTK specification I think, where the `QuadraticQuad` element is 8 nodes.

File: src/fe_values.jl (hunk 3 above)
Comment (KristofferC) [inline src/fe_values.jl]: yes

**Comment by fredrikekre** [comment]
What about the names? :)

**Comment by fredrikekre** [comment]
YOLO.

**Comment by KristofferC** [comment]
Pff, deletar sin kommentar och grejjer,

**Pull request merged**